### PR TITLE
Add platform checkout prototype

### PR DESCRIPTION
**Pull request by waclawjacek**
#### Changes proposed in this Pull Request

Adds platform checkout support prototype.

The PR introduces the `_wcpay_feature_platform_checkout` feature flag which, when enabled, will register a prototype of functionality related to platform checkout.

Note that in its current state the prototype is not safe for production use.

#### Testing instructions

1. Set the `_wcpay_feature_platform_checkout` option to (int) `1`.
2. The option to checkout via platform checkout should now appear in cart and checkout blocks.
3. Set the `_wcpay_feature_platform_checkout` option to (int) `0`.
4. The option to checkout via platform checkout should **not** appear in cart and checkout blocks.
5. Also verify that it is not possible to access REST endpoints without a valid nonce, the `wc_ajax_wcpay_init_platform_checkout` AJAX action, and that the `X-Wcpay-Platform-Checkout-User` header has no effect when the feature flag is disabled (and when enabled, all of these work).

-------------------

- [ ] Added changelog entries to both `readme.txt` and `changelog.txt` (or does not apply)
   - Does not apply - prototype behind a feature flag.
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)
   - Does not apply - prototype behind a feature flag.

**Post merge**

- [ ] Added testing instructions to the [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) (or does not apply)
   - Does not apply - prototype behind a feature flag.

### CODE REVIEW

File: includes/class-wc-payments.php [hunk 1]
```diff
@@ -771,4 +773,77 @@ public static function remove_woo_admin_notes() {
...
+	/**
+	 * Determine the current user
+	 *
+	 * @param WP_User|int $user The user to determine.
+	 */
+	public static function determine_current_user_for_platform_checkout( $user ) {
+		if ( $user ) {
+			return $user;
+		}
+		$headers = getallheaders();
```
Comment (superdav42) [inline includes/class-wc-payments.php:835]: Not related to renaming this function but getallheaders() is not the function we should use here. It is only available when running in apache or nginx + php-fpm 7.3+ and not in the cli. I ran into this problem running this command:
```
wp option set _wcpay_feature_platform_checkout 0
```
We should update it to use `$_SERVER['HTTP_X-WCPAY-PLATFORM-CHECKOUT-USER']` which will always work and not cause a problem in cli.

File: includes/class-wc-payments.php [hunk 2]
```diff
@@ -771,4 +773,77 @@ public static function remove_woo_admin_notes() {
...
+		if ( ! isset( $headers['X-Wcpay-Platform-Checkout-User'] ) || ! is_numeric( $headers['X-Wcpay-Platform-Checkout-User'] ) ) {
```
Comment (superdav42) [inline includes/class-wc-payments.php:836]:
```suggestion
		if ( ! isset( $_SERVER['HTTP_X-WCPAY-PLATFORM-CHECKOUT-USER'] ) || ! is_numeric( $headers['X-Wcpay-Platform-Checkout-User'] ) ) {
```

File: includes/class-wc-payments.php [hunk 3]
```diff
@@ -771,4 +773,77 @@ public static function remove_woo_admin_notes() {
...
+		}
+
+		add_filter(
+			'woocommerce_cookie',
+			function( string $cookie_hash ) {
+				return 'platform_checkout_session';
+			}
+		);
+
+		return (int) $headers['X-Wcpay-Platform-Checkout-User'];
```
Comment (superdav42) [inline includes/class-wc-payments.php:847]:
```suggestion
		return (int) $headers['X-Wcpay-Platform-Checkout-User'];
```

**Review by superdav42 — CHANGES_REQUESTED** [review]
Code works great and test look good. Only ran into one problem but from the original code. I think it's a big enough problem we should just fix it now rather than a separate PR.

File: includes/class-wc-payments.php (hunk 3 above)
Comment (reykjalin) [reply includes/class-wc-payments.php:847]: @superdav42 I think the changes you proposed here were meant to be the same changes you proposed in https://github.com/Automattic/woocommerce-payments/pull/3481#discussion_r764266807 ?

So should be the following instead I think?

```suggestion
		return (int) $_SERVER['HTTP_X-WCPAY-PLATFORM-CHECKOUT-USER'];
```

File: includes/class-wc-payments.php (hunk 2 above)
Comment (reykjalin) [reply includes/class-wc-payments.php:836]: Also, shouldn't both parts of the if-statement be changed to use `$_SERVER['HTTP_X-WCPAY-PLATFORM-CHECKOUT-USER']`? 🤔

**Review by reykjalin — APPROVED** [review]
LGTM 🎉 

Let's just address the comments from @superdav42's review before we merge 🙂

File: includes/class-wc-payments.php (hunk 2 above)
Comment (superdav42) [reply includes/class-wc-payments.php:836]: Yes

File: includes/class-wc-payments.php (hunk 3 above)
Comment (superdav42) [reply includes/class-wc-payments.php:847]: Yes

File: includes/class-wc-payments.php (hunk 1 above)
Comment (waclawjacek) [reply includes/class-wc-payments.php]: Thanks for pointing this out! I have tested this and noticed the header arrives in the `$_SERVER` superglobal as: `HTTP_X_WCPAY_PLATFORM_CHECKOUT_USER`. I have used that for the key here and in all of the other places.

**Review by superdav42 — APPROVED** [review]
Thanks

File: includes/class-wc-payments.php (hunk 3 above)
Comment (khag7) [reply includes/class-wc-payments.php]: Uhhh this seems ripe for exploitation

**Pull request merged**